util/src/main/java/com/google/cloud/hadoop/util/RetryHttpInitializer.java

```diff
@@ -60,119 +58,124 @@
-        switch (httpResponse.getStatusCode()) {
-          case HTTP_SC_TOO_MANY_REQUESTS:
-            logger.atInfo().atMostEvery(10, SECONDS).log(
-                LOG_MESSAGE_FORMAT,
-                httpResponse.getStatusCode(),
-                httpRequest.getRequestMethod(),
-                httpRequest.getUrl());
-            break;
-          default:
-            logger.atInfo().atMostEvery(10, SECONDS).log(
-                "Encountered status code %d (and maybe others) when sending %s request to URL '%s'."
-                    + " Delegating to response handler for possible retry.",
-                httpResponse.getStatusCode(), httpRequest.getRequestMethod(), httpRequest.getUrl());
-        }
-      } else if (responseCodesToLog.contains(httpResponse.getStatusCode())) {
-        logger.atInfo().log(
-            LOG_MESSAGE_FORMAT,
-            httpResponse.getStatusCode(),
-            httpRequest.getRequestMethod(),
-            httpRequest.getUrl());
-      }
+  /**
+   * @param credentials A credentials which will be used to initialize on HttpRequests and as the
+   *     delegate for a {@link UnsuccessfulResponseHandler}.
+   * @param options An options that configure {@link RetryHttpInitializer} instance behaviour.
+   */
+  public RetryHttpInitializer(Credentials credentials, RetryHttpInitializerOptions options) {
+    this.credentials = credentials == null ? null : new HttpCredentialsAdapter(credentials);
+    this.options = options;
+  }
 
-      return delegateResponseHandler.handleResponse(httpRequest, httpResponse, supportsRetry);
+  @Override
+  public void initialize(HttpRequest request) throws IOException {
+    // Initialize request with credentials and let CredentialsOrBackoffResponseHandler
+    // to refresh credentials later if necessary
+    if (credentials != null) {
+      credentials.initialize(request);
     }
 
-    @Override
-    public boolean handleIOException(HttpRequest httpRequest, boolean supportsRetry)
-        throws IOException {
-      // We sadly don't get anything helpful to see if this is something we want to log. As a result
-      // we'll turn down the logging level to debug.
-      logger.atFine().log("Encountered an IOException when accessing URL %s", httpRequest.getUrl());
-      return delegateIOExceptionHandler.handleIOException(httpRequest, supportsRetry);
+    request
+        // Request will be retried if server errors (5XX) or I/O errors are encountered.
+        .setNumberOfRetries(options.getMaxRequestRetries())
+        // Set the timeout configurations.
+        .setConnectTimeout(toIntExact(options.getConnectTimeout().toMillis()))
+        .setReadTimeout(toIntExact(options.getReadTimeout().toMillis()))
+        // Supply a new handler for unsuccessful return codes. 401 Unauthorized will be handled by
+        // the Credentials, 410 Gone will be logged, and 5XX will be handled by a backoff handler.
```

Comment:
nit: can we move this comment to javadoc of the handler class ?

Reply:
Done.